Scanner engine improvementsv2

thenenadx opened the pull request:
* Refactoring of the scanner to support loading of different rules
* Modified scanner tests
* Some tests have been disabled because they need to be rewritten by a bit

blueandgold commented on google/cloud/security/scanner/scanner.py:
```diff
@@ -84,60 +121,112 @@ def main(_):
-    if not org_policies and not project_policies:
-        LOGGER.warn('No policies found. Exiting.')
-        sys.exit()
+    data_pipeline_class, _ = _create_rules_engine(
```
As noted above, perhaps we can create a map and not need the  ```_create_rules_engine()```?

For example, we can do something like this:
```python

scanner_map  = {
    'IamPolicy': scanners.IamPolicyScanner,
    'Foo': scanners.FooScanner,
    'Bar':  scanners.BarScanner,
  }

# So, that we can do something like this:
scanner = scanner_map['IamPolicy'](snapshot_timestamp)
```

blueandgold commented on google/cloud/security/scanner/scanner.py:
```diff
@@ -69,13 +69,50 @@
+        rules_engine_filename = FLAGS.use_engine
+
+    # Load rules engine according to command line flag
+    rules_engine_class, rules_engine_name = _create_rules_engine(
```
Hmmm, perhaps we can create a map and not need the _create_rules_engine()?

For example:
```python
ENGINE_TO_DATA_MAP =  {
    'IamRulesEngine': audit.IamRulesEngine,
    'FooRulesEngine': audit.FooRulesEngine,
    'BarRulesEngine': audit.BarRulesEngine,
}

# So, that we can do something like this:
rules_engine = ENGINE_TO_DATA_MAP['IamRulesEngine'](rules_file_path=FLAGS.rules)
```

blueandgold commented on google/cloud/security/scanner/pipelines/base_data_pipeline.py:
```diff
@@ -0,0 +1,47 @@
+LOGGER = log_util.get_logger(__name__)
+
+
+class BaseDataPipeline(object):
```
Pending carise's comment, but perhaps we can have more appropriate naming here?

Instead of ```pipeline```, how about ```scanner```? 
And should use just drop ```data```?
How does this look?
```
google/cloud/security/scanner/scanners
google/cloud/security/scanner/scanners/base_scanner.py
google/cloud/security/scanner/scanners/iam_rules_scanner.py
```

blueandgold commented on google/cloud/security/scanner/pipelines/load_iam_rules_engine_pipeline.py:
```diff
@@ -0,0 +1,125 @@
+LOGGER = log_util.get_logger(__name__)
+
+
+class LoadIamDataPipeline(base_data_pipeline.BaseDataPipeline):
```
nit: naming :)  See my comment above for proposal.

thenenadx commented on google/cloud/security/scanner/scanner.py:
```diff
@@ -69,13 +69,50 @@
 
+flags.DEFINE_bool('list_engines', False, 'List all rule engines')
+
+flags.DEFINE_string('use_engine', None, 'Which engine to use')
```
Yep, so either it should be scripted to go through all the engines or we can improve it even more to go through the whole map.

thenenadx commented:
@blueandgold The latest commit addresses all your feedback, except the one from the comment from above :) It's very easy to add if we feel like going through all the rules

carise commented on tests/scanner/scanner_test.py:
```diff
@@ -32,6 +32,7 @@
 from google.cloud.security.scanner import scanner
 from google.cloud.security.scanner.audit import iam_rules_engine as ire
 from google.cloud.security.scanner.audit import rules as audit_rules
+from google.cloud.security.scanner.scanners import iam_rules_scanner as irs
```
should we separate out the individual scanner tests into its own test suite and keep the scanner runner tests here?

thenenadx replied:
Yep, that was my plan as part of fixing all tests, especially the ones I've disabled.

blueandgold commented:
Adding `cla: yes` label, as the PR author has ack'd signing the CLA in PR # 240.
